// dice/dice.js
#!/usr/bin/env node
/*jslint nodejs:true */

var
	http = require("http"),
	url = require("url"),
	argv = require("optimist").string("h", "s").argv;

// Input: String in the form ndfaces or nDfaces. E.g. "2d6".
// Output: n random rolls from [1, faces]. E.g. [3, 4].

function roll(s) {
	var
		roll = s.toLowerCase() || "2d6",
		parts = roll.split("d"),
		n = parseInt(parts[0], 10) || 2,
		faces = parseInt(parts[1], 10) || 6;

	var rs = [];

	for (var i = 0; i < n; i++) {
		var r = Math.floor(Math.random() * faces) + 1;
		rs.push(r);
	}

	return rs;
}

exports.roll = roll;

function sum(array) {
	var total = 0;

	for (var i in array) {
		total += array[i];
	}

	return total;
}

exports.sum = sum;

function page(req) {
	var
		query = url.parse(req.url, true).query,
		r = query.roll || "2d6",
		rs = roll(r),
		total = sum(rs);

	return "<!DOCTYPE html>" +
		"<head>" +
		"<title>" + r + "</title><meta http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\" />" +
		"</head>" +
		"<body onload=\"document.getElementById('roll').focus();\">" +
		"<style type=\"text/css\" />" +
		"body, input { text-align: center; }" +
		"input { font-size: 150%; }" +
		"a { color: #000; }" +
		"a:visited { color: #000; }" +
		"#logo { font-size: 500%; }" +
		"#dice_form { display: block; margin: 0 auto; width: 400px; }" +
		"label { display: block; float: left; width: 5%; padding-right: 1%; }" +
		"#github { position: fixed; bottom: 1em; right: 1em; }" +
		"</style>" +
		"<h1 id=\"logo\"><a href=\"/\">D&D Dice</a></h1>" +
		"<div id=\"dice_form\">" +

		"<form action=\"\" method=\"get\">" +

		"<input type=\"text\" id=\"r\" name=\"roll\" size=\"10\" value=\"" + r + "\" />" +

		" " +

		"<input type=\"submit\" value=\"Roll\" id=\"roll\">" +

		"<p><label>Rolls</label>" + rs + "<br/>" +
		"<label>Sum</label>" + total + "</p>" +

		"</div>" +
		"<div id=\"github\"><a href=\"https://github.com/mcandre/node-dice\">GitHub</a></div>" +
		"</body></html";
}

function server() {
	http.createServer(function (req, res) {
		res.writeHead(200, {"Content-Type": "text/html"});
		res.end(page(req));
	}).listen(5125, "localhost");

	console.log("Server running at http://localhost:5125/");
}

exports.server = server;

function commandLine() {
	process.stdin.resume();
	process.stdin.setEncoding("UTF8");

	process.stdin.on("data", function(chunk) {
		var rs = roll(chunk);
		console.log(rs);
		console.log(sum(rs));
	});

	process.stdin.on("end", function() {});
}

exports.commandLine = commandLine;

function usage() {
	console.log("dice.js [options]");
	console.log("-s\tServer");
	console.log("-h\tUsage");
}

exports.usage = usage;

function main() {
	if ("h" in argv) {
		usage();
	}
	else if ("s" in argv) {
		server();
	}
	else {
		commandLine();
	}
}

if (!module.parent) { main(); }
